File: src/onion/compiler/env/LocalFrame.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.env;

import java.util.*;

import org.onion_lang.onion.lang.core.type.TypeRef;


/**
 * @author Kota Mizushima
 * Date: 2005/06/28
 */
public class LocalFrame {
  private LocalFrame parent;
  private LocalScope scope;
  private List<LocalScope> allScopes;
  private int maxIndex;
  private boolean closed;
  
  /**
   * Creates a new frame.
   * @param parent the parent frame
   */
  public LocalFrame(LocalFrame parent) {
    this.parent = parent;
    this.scope = new LocalScope(null);
    this.allScopes = new ArrayList<LocalScope>();
    allScopes.add(scope);
  }
  
  /**
   * Gets the parent frame.
   * @return
   */
  public LocalFrame getParent(){
    return parent;
  }
  
  /**
   * Opens a new scope.
   */
  public void openScope(){
    scope = new LocalScope(scope);
    allScopes.add(scope);
  }
  
  /**
   * Closes the current scope.
   */
  public void closeScope(){
    scope = scope.getParent();
  }

  LocalScope getScope(){
    return scope;
  } 
  
  public LocalBinding[] entries(){
    Set<LocalBinding> entries = entrySet();
    LocalBinding[] binds = new LocalBinding[entries.size()];
    Iterator<LocalBinding> iterator = entries.iterator();
    for(int i = 0; i < binds.length; i++){
      binds[i] = iterator.next();
    }
    Arrays.sort(binds, new Comparator<LocalBinding>(){
      public int compare(LocalBinding b1, LocalBinding b2) {
        int i1 = b1.getIndex(), i2 = b2.getIndex();
        return i1 < i2 ? -1 : i1 > i2 ? 1 : 0;
      }
    });
    return binds;
  }
  
  public int addEntry(String name, TypeRef type){
    LocalBinding bind = scope.get(name);    
    //if name is already registered, it returns -1 which means failure.
    if(bind != null) return -1;
    
    int index = maxIndex;
    maxIndex++;
    scope.put(name, new LocalBinding(index, type));
    return index;
  }
  
  public ClosureLocalBinding lookup(String name){
    LocalFrame frame = this;
    int frameIndex = 0;
    while(frame != null){
      LocalBinding binding = frame.scope.lookup(name);
      if(binding != null){
        return new ClosureLocalBinding(
          frameIndex, binding.getIndex(), binding.getType()
        );
      }      
      frameIndex++;
      frame = frame.parent;
    }
    return null;
  }

  public ClosureLocalBinding lookupOnlyCurrentScope(String name) {
    LocalBinding binding = scope.get(name);
    if(binding != null){
      return new ClosureLocalBinding(
        0, binding.getIndex(), binding.getType()
      );
    }
    return null;
  }
  
  public void setAllClosed(boolean closed){
    LocalFrame frame = this;
    while(frame != null){
      frame.setClosed(closed);
      frame = frame.getParent();
    }
  }
  
  public void setClosed(boolean closed){
    this.closed = closed;
  }
  
  public boolean isClosed(){
    return closed;
  }
  
  public int depth(){
    LocalFrame frame = this;
    int depth = -1;
    while(frame != null){
      depth++;
      frame = frame.getParent();
    }
    return depth;
  }
  
  private Set<LocalBinding> entrySet(){
    Set<LocalBinding> entries = new HashSet<LocalBinding>();
    for(LocalScope s : allScopes) {
      entries.addAll(s.entries());
    }
    return entries;
  }
}

File: src/main/scala/onion/compiler/LocalFrame.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler

import java.util.{Arrays => JArrays}
import scala.collection.mutable

/**
 * @author Kota Mizushima
 *
 */
class LocalFrame(val parent: LocalFrame) {
  var scope = new LocalScope(null)
  var closed: Boolean = false

  private val allScopes: mutable.Buffer[LocalScope] = mutable.Buffer[LocalScope]()
  private var maxIndex: Int = 0

  allScopes += scope

  def open[A](block: => A): A = {
    try {
      scope = new LocalScope(scope)
      allScopes += scope
      block
    } finally {
      scope = scope.parent
    }
  }

  def entries: Seq[LocalBinding] = {
    val binds: Array[LocalBinding] = entrySet.toArray
    JArrays.sort(binds, (b1: LocalBinding, b2: LocalBinding) => {
      val i1 = b1.index
      val i2 = b2.index
      if (i1 < i2) -1 else if (i1 > i2) 1 else 0
    })
    binds.toSeq
  }

  def add(name: String, `type` : TypedAST.Type): Int = {
    add(name, `type`, isMutable = true, isBoxed = false)
  }

  def add(name: String, `type`: TypedAST.Type, isMutable: Boolean): Int = {
    add(name, `type`, isMutable, isBoxed = false)
  }

  def add(name: String, `type`: TypedAST.Type, isMutable: Boolean, isBoxed: Boolean): Int = {
    val bind = scope.get(name)
    bind match {
      case Some(_) => -1
      case None =>
        val index = maxIndex
        maxIndex += 1
        scope.put(name, LocalBinding(index, `type`, isMutable, isBoxed))
        index
    }
  }

  /** Iterator over frame hierarchy starting from this frame */
  private def frames: Iterator[LocalFrame] =
    Iterator.iterate(this)(_.parent).takeWhile(_ != null)

  /** Tail-recursive lookup through frame hierarchy */
  @scala.annotation.tailrec
  private def lookupInFrames(frame: LocalFrame, frameIndex: Int, name: String): ClosureLocalBinding =
    if (frame == null) null
    else frame.scope.lookup(name) match {
      case null => lookupInFrames(frame.parent, frameIndex + 1, name)
      case binding => new ClosureLocalBinding(frameIndex, binding.index, binding.tp, binding.isMutable, binding.isBoxed)
    }

  def lookup(name: String): ClosureLocalBinding = lookupInFrames(this, 0, name)

  def lookupOnlyCurrentScope(name: String): ClosureLocalBinding = {
    scope.get(name).map { binding =>
      new ClosureLocalBinding(0, binding.index, binding.tp, binding.isMutable, binding.isBoxed)
    }.orNull
  }

  /** Gets all variable names visible from the current scope (including parent frames) */
  def allNames: Set[String] = frames.flatMap(_.scope.allNames).toSet

  def setAllClosed(closed: Boolean): Unit = frames.foreach(_.closed = closed)

  def depth: Int = frames.length - 1

  private def entrySet: mutable.Set[LocalBinding] = {
    val entries = new mutable.HashSet[LocalBinding]()
    mutable.Set[LocalBinding]() ++ (for (s1 <- allScopes; s2 <- s1.entries) yield s2)
  }

}
